File: leetcode/Java/0124-h-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private class ResultType {
        int root2any, any2any;
        ResultType(int root2any, int any2any) {
            this.root2any = root2any;
            this.any2any = any2any;
        }
    }

    private ResultType helper(TreeNode root) {
        if (root == null)
            return new ResultType(Integer.MIN_VALUE, Integer.MIN_VALUE);
        
        ResultType left = helper(root.left);
        ResultType right = helper(root.right);

        int root2any = Math.max(Math.max(left.root2any, right.root2any), 0) + root.val;

        int any2any = Math.max(left.any2any, right.any2any);

        any2any = Math.max(
            any2any,
            Math.max(left.root2any, 0) + root.val + Math.max(right.root2any, 0)
        );

        return new ResultType(root2any, any2any);
    }

    public int maxPathSum(TreeNode root) {
        ResultType res = helper(root);
        return res.any2any;
    }
}

File: leetcode/C++/0124-h-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int max_sum = INT_MIN;
public:
    int maxPathSum(TreeNode* root) {
        max_gain(root);
        return max_sum;
    }
    
    int max_gain(TreeNode* node) {
        if (!node) return 0;
        int left_gain = max(max_gain(node->left), 0);
        int right_gain = max(max_gain(node->right), 0);
        
        int price_newpath = node->val + left_gain + right_gain;
        
        max_sum = max(max_sum, price_newpath);
        
        return node->val + max(left_gain, right_gain);
    }
};